### ISSUE TITLE
Add PHP support 

### ISSUE THREAD

**Issue by kcloze** [pyroscope-io/pyroscope#30]
Beautiful interface, welcome to surpport PHP.

**Comment by markhuot** [comment]
I would like to help with this but am not sure where to start. I'm thinking we could leverage some of the existing sampling profilers out there but am not sure what changes need to be made to their output for Pyroscope to read them correctly.

Examples,

- https://github.com/adsr/phpspy
- https://github.com/sj-i/php-profiler
- https://github.com/nikic/sample_prof

**Comment by petethepig** [comment]
@markhuot This is already very useful! `phpspy` looks very promising. Profilers we use for python and ruby (`py-spy` and `rbspy`) use a similar approach for getting the profiling data (they look at process memory X number of times per second) and I bet it would be relatively straightforward to add `phpspy` to pyroscope if we modify it just a little bit.

---

For example, here's how we integrated rbspy into pyroscope:

* we modified rbspy to compile into a library instead of an executable binary
* we modified rbspy code to implement the following interface: https://github.com/pyroscope-io/rbspy/blob/main/lib/rbspy.h#L3-L5
```
int rbspy_init(pid_t pid, void* err_ptr, int err_len);
int rbspy_cleanup(pid_t pid, void* err_ptr, int err_len);
int rbspy_snapshot(pid_t pid, void* ptr, int len, void* err_ptr, int err_len);
```
* once you can build a static library that complies with this interface, integrating it with go is pretty straightforward: https://github.com/pyroscope-io/pyroscope/blob/main/pkg/agent/rbspy/rbspy.go#L1-L74

---

I hope this is enough information to get started. @markhuot Ping me in our [Slack](https://pyroscope.io/slack), I would love to help, this seems very promising.

**Comment by markhuot** [comment]
Thanks for these pointers @petethepig! I've dug in to `phpspy` and have a much better understanding of how it's working and what it's doing now. Unfortunately the C work is a bit outside my day-to-day experience so I'm getting up to speed on that. In your first bullet you say "we modified rbspy to compile into a library instead of an executable binary." Is that something you did inside the `rbspy` logic or a flag you flip on the compiler?

**Comment by markhuot** [comment]
Ah, I think I see. I've been looking through this commit for inspiration, https://github.com/pyroscope-io/rbspy/commit/00dbbe4acd9114d13c16de8a3c2f134be59af8e3

**Comment by petethepig** [comment]
@markhuot I think for a C library it will be more of a flag in the compiler type thing.

**Comment by adrianandreias** [comment]
PHP support would be really useful.

**Comment by petethepig** [comment]
Started working on it in #157

**Comment by petethepig** [comment]
Also made this diagram describing the relationship between components: ![agents](https://user-images.githubusercontent.com/662636/115311366-69bfbc80-a124-11eb-9891-8f94f9df0eac.png)

**Comment by petethepig** [comment]
@adrianandreias @arrack @jamesaspence @jnlin @kcloze @knibals @Kortivex @markhuot @Slach @tooolbox @topazas @vearutop 

We just released [v0.0.31](https://github.com/pyroscope-io/pyroscope/releases/tag/v0.0.31) with php support. Please try it out, we would love to hear feedback from any of you.

Currently it only works on Linux and the way to start it is by using `pyroscope exec`:
```
pyroscope exec php main.php
```

Or you can also connect to an existing php process:
```
pyroscope connect -pid <pid of the process>
```

We also have a dockerized example project [here](https://github.com/pyroscope-io/pyroscope/tree/main/examples/php)

**Comment by abeaumont** [comment]
Closing this, PHP support is already provided through phpspy.
